[FLINK-36696] [flink-autoscaler-plugin-jdbc] Switch sql connection usages to datasource

sharath1709 opened the pull request:
<!--
*Thank you very much for contributing to the Apache Flink Kubernetes Operator - we are happy that you want to help us improve the project. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix][docs] Fix typo in event time introduction` or `[hotfix][javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can read more on how we use GitHub Actions for CI [here](https://nightlies.apache.org/flink/flink-kubernetes-operator-docs-main/docs/development/guide/#cicd).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

This pull request makes changes to replace the usage of SQL connection with DataSource to improve the multithreaded performance of the JDBC plugin.


## Brief change log

  - *Replace all usages of SQL connection with DataSource *

## Verifying this change
<!--
Please make sure both new and modified tests in this PR follows the conventions defined in our code quality guide: https://flink.apache.org/contributing/code-style-and-quality-common.html#testing
-->

This change is already covered by existing unit tests. The test cases have been modified appropriately to use DataSource rather than Connection everywhere


## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changes to the `CustomResourceDescriptors`: no
  - Core observer or reconciler logic that is regularly executed: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

1996fanrui commented on flink-autoscaler-plugin-jdbc/src/test/java/org/apache/flink/autoscaler/jdbc/state/AbstractJdbcStateStoreITCase.java:
```diff
@@ -45,24 +45,17 @@
         this.jdbcStateStore = new JdbcStateStore(jdbcStateInteractor);
     }
 
-    @AfterEach
-    void afterEach() throws SQLException {
-        if (conn != null) {
-            conn.close();
-        }
-    }
```
Why don't close dataSource here?

As I understand, each test will create a new dataSource. The old dataSource will leak some connections if we don't close dataSource here.

1996fanrui commented on flink-autoscaler-plugin-jdbc/src/test/java/org/apache/flink/autoscaler/jdbc/state/AbstractJdbcStateInteractorITCase.java:
```diff
@@ -42,33 +42,32 @@ void testAllOperations() throws Exception {
-        try (var conn = getConnection()) {
-            var jdbcStateInteractor = new JdbcStateInteractor(conn);
-            assertThat(jdbcStateInteractor.queryData(jobKey)).isEmpty();
+        var dataSource = getDataSource();
```
It's similar with the last comment, we didn't close the dataSource.

1996fanrui reviewed (commented):
Thanks @sharath1709 for the contribution!

Overall LGTM! And I still have some comments:

1. The first question is related to connection leak for testing. This PR will create lots of new dataSources, but the close isn't called. IIUC, it will have connection leak.
2. The CI is failed, please take a look in your free time.
3. We encountered a bug related to this fix recently. The database master is crashed, the slave is promoted to be the master. The jdbc state store and event handler won't work due to they are using the fixed connection. 
    - As I understand,  DataSource could solve this problem as well. 
    - When old connection is crashed, DataSource will create a new physical connection to connect the new database master server.
    - Would you mind adding a test to check it? In the ITCase, we only have one databse server, I'm not sure could it be reproduced when the database server is restarted.

sharath1709 commented:
Thanks a lot @1996fanrui for the quick review. Please find my replies below

1. Typically, DataSource connections will be relinquished automatically after a time out and DataSource isn't Closeable. However, some DataSource implementations like HikariDataSource implement the Closeable interface to close the ConnectionPool. We can add a check to see if the DataSource is Closeable and then close it. Additionally, I've attempted to make the classes that hold the datasource objects Autocloseable as well so that they could be created in a try-with-resources block
2. Thanks, I've tried to fix most of the issues but might need a couple more iterations.
3. Great point, I will try to add a test case for this scenario soon

1996fanrui commented on flink-autoscaler-plugin-jdbc/src/test/java/org/apache/flink/autoscaler/jdbc/testutils/databases/derby/DerbyExtension.java:
```diff
@@ -34,8 +37,10 @@ public class DerbyExtension implements BeforeAllCallback, AfterAllCallback, Afte
 
-    public Connection getConnection() throws Exception {
-        return DriverManager.getConnection(JDBC_URL);
+    public DataSource getDataSource() {
```
Similar with the last comment.

1996fanrui commented on flink-autoscaler-standalone/src/test/java/org/apache/flink/autoscaler/standalone/AutoscalerStateStoreFactoryTest.java:
```diff
@@ -67,14 +66,12 @@ void testCreateJdbcStateStore() throws Exception {
 
         try {
             conf.set(JDBC_URL, String.format("%s;shutdown=true", jdbcUrl));
-            HikariJDBCUtil.getConnection(conf).close();
```
IIUC, `conf.set(JDBC_URL, String.format("%s;shutdown=true", jdbcUrl))` means shutdown the derby server. Don't we need to call it here?

sharath1709 replied:
I see, I wasn't familiar with this and assumed it only closes the connection. We can continue to keep the same logic.

1996fanrui commented on flink-autoscaler-standalone/src/test/java/org/apache/flink/autoscaler/standalone/AutoscalerEventHandlerFactoryTest.java:
```diff
@@ -68,14 +67,12 @@ void testCreateJdbcEventHandler() throws Exception {
 
         try {
             conf.set(JDBC_URL, String.format("%s;shutdown=true", jdbcUrl));
-            HikariJDBCUtil.getConnection(conf).close();
```
Same with the last comment.

1996fanrui commented on flink-autoscaler-plugin-jdbc/src/test/java/org/apache/flink/autoscaler/jdbc/event/AbstractJdbcAutoscalerEventHandlerITCase.java:
```diff
@@ -459,7 +462,8 @@ void testDeleteCounterWhenIdNotConsecutive() throws Exception {
                         .orElseThrow();
 
-        try (Connection connection = getConnection();
+        var dataSource = getDataSource();
```
This temporary data source isn't closed. 

Or could we reuse the dataSource of `beforeEach` instead of creating a new one?

1996fanrui commented on flink-autoscaler-plugin-jdbc/src/test/java/org/apache/flink/autoscaler/jdbc/event/AbstractJdbcEventInteractorITCase.java:
```diff
@@ -46,63 +46,63 @@ void testAllOperations() throws Exception {
 
         // The datetime precision is seconds in MySQL by default.
         var createTime = Instant.now().truncatedTo(ChronoUnit.SECONDS);
-        try (var conn = getConnection()) {
-            var jdbcEventInteractor = new JdbcEventInteractor(conn);
```
```
        try(var jdbcEventInteractor = new JdbcEventInteractor(getDataSource())) {
            
        }
```

It maybe a simpler choice, we don't need to check `dataSource` and close it in the end.

sharath1709 replied:
Mb. Agree.

1996fanrui commented on flink-autoscaler-plugin-jdbc/src/test/java/org/apache/flink/autoscaler/jdbc/state/AbstractJdbcStateInteractorITCase.java:
```diff
@@ -42,32 +42,35 @@ void testAllOperations() throws Exception {
         var value1 = "value1";
         var value2 = "value2";
         var value3 = "value3";
-        try (var conn = getConnection()) {
-            var jdbcStateInteractor = new JdbcStateInteractor(conn);
```
```
        try(var jdbcStateInteractor = new JdbcStateInteractor(getDataSource())) {
            
        }
```

It maybe a simpler choice, we don't need to check `dataSource` and close it in the end.

sharath1709 replied:
updated

1996fanrui commented on flink-autoscaler-plugin-jdbc/src/test/java/org/apache/flink/autoscaler/jdbc/state/JobStateViewTest.java:
```diff
@@ -36,25 +35,18 @@
-    @AfterEach
-    void afterEach() throws SQLException {
-        if (conn != null) {
-            conn.close();
```
Why don't close `jdbcStateInteractor` here?

sharath1709 replied:
Thanks, updated.

1996fanrui reviewed (commented):
Hi @sharath1709 , thanks for the update!

I left some comments, and the CI is still failed (It cannot be compiled for now. ). 

You could try to execute ` mvn  clean install -DskipTests` in your local before next push, thanks~

1996fanrui commented on flink-autoscaler-standalone/src/test/java/org/apache/flink/autoscaler/standalone/AutoscalerEventHandlerFactoryTest.java:
```diff
@@ -68,14 +68,14 @@ void testCreateJdbcEventHandler() throws Exception {
-        try {
-            conf.set(JDBC_URL, String.format("%s;shutdown=true", jdbcUrl));
-            HikariJDBCUtil.getConnection(conf).close();
+        try (var datasource = HikariJDBCUtil.getDataSource(conf)) {
```
It's similar with the last comment.

sharath1709 replied:
good catch, fixed

1996fanrui reviewed (commented):
Thanks @sharath1709 for the update, overall LGTM. I left 2 minor comments, please take a look in your free time, thanks~

> 3. We encountered a bug related to this fix recently. The database master is crashed, the slave is promoted to be the master. The jdbc state store and event handler won't work due to they are using the fixed connection.
> 
>     - As I understand, DataSource could solve this problem as well.
>     - When old connection is crashed, DataSource will create a new physical connection > to connect the new database master server.
>     - Would you mind adding a test to check it? In the ITCase, we only have one databse server, I'm not sure could it be reproduced when the database server is restarted.

I have checked it manually on my Mac, it works well with this PR. Also, as we discussed before, it's better to add a test to check it.

sharath1709 commented:
Thanks @1996fanrui for patience with the review. I addressed the remaining comments and added a test case for db restart. I also verified that the test case fails without the change in this PR.

1996fanrui commented:
> I addressed the remaining comments and added a test case for db restart to `AbstractJdbcStateInteractorITCase`. The same could be added to `AbstractJdbcEventInteractorITCase` if needed but it's a bit redundant.

Sounds make sense to me, testing restart in AbstractJdbcStateInteractorITCase is enough, but the test is failed. 
https://github.com/apache/flink-kubernetes-operator/actions/runs/12939940745/job/36097810802?pr=929#step:8:1025

Also, I may need to do another round of overall review after the test passes.

sharath1709 commented:
@1996fanrui It appears that MySQL/Postgresql DB restart test cases are failing in CI with SQLTransientException. Let me update the test to ignore transient exception as typically subsequent connection should be successful. We may also consider adding retries but I feel it's overkill. lmk your thoughts

1996fanrui commented:
> @1996fanrui It appears that MySQL/Postgresql DB restart test cases are failing in CI with SQLTransientException. Let me update the test to ignore transient exception as typically subsequent connection should be successful. Before this PR, the test case produces SQLNonTransientConnectionException. We may also consider adding retries but I feel it's overkill. lmk your thoughts

After I debug it on my Mac, I don't think catch `SQLTransientException` is reasonable.  Please check the following figure, the `bbbbb` is not executed. It means we cannot ensure the connection works after restarting. So I think retry mechanism is needed, we need to retry queryData from data base until the result is expected.

Also, I added the retry mechanism on my Mac, but it still doesn't work, I found after restart, the port of MySQL Container is changed. It means the original data source won't be used.

![](https://github.com/user-attachments/assets/e48e9837-b2b6-4e7c-86f4-26080ae891b2)

sharath1709 commented:
That sounds great, let me remove the test. I'm not the expert on this situation either and ChatGPT originally suggested starting and stopping the container to simulate a restart. I will try to explore it in my free time and add the test case later

1996fanrui commented:
One test fails since ` too many clients already`, I'm not sure is there any client or connection leak?

```
Error:  Tests run: 17, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 33.745 s <<< FAILURE! - in org.apache.flink.autoscaler.jdbc.event.PostgreSQLJdbcAutoscalerEventHandlerITCase
Error:  org.apache.flink.autoscaler.jdbc.event.PostgreSQLJdbcAutoscalerEventHandlerITCase.testEventIntervalWithoutMessageKey  Time elapsed: 1.024 s  <<< ERROR!
com.zaxxer.hikari.pool.HikariPool$PoolInitializationException: Failed to initialize pool: FATAL: sorry, too many clients already
Caused by: org.postgresql.util.PSQLException: FATAL: sorry, too many clients already
```


https://github.com/apache/flink-kubernetes-operator/actions/runs/12958019250/job/36164630541?pr=929#step:8:2293

1996fanrui commented on flink-autoscaler-plugin-jdbc/src/test/java/org/apache/flink/autoscaler/jdbc/testutils/databases/postgres/PostgreSQLExtension.java:
```diff
@@ -66,7 +70,8 @@ public void afterAll(ExtensionContext extensionContext) {
     public void afterEach(ExtensionContext extensionContext) throws Exception {
-        try (var conn = getConnection();
+        try (var dataSource = getDataSource();
+                var conn = getDataSource().getConnection();
```
```suggestion
                var conn = dataSource.getConnection();
```

@sharath1709 I found a bug here, it should use data source instead of ` getDataSource()` here.

The second ` getDataSource()` is not closed. That's why connection is leaked.

https://github.com/apache/flink-kubernetes-operator/pull/929#issuecomment-2614195894

1996fanrui reviewed (approved):
Thanks @sharath1709 for the patience!

LGTM. I will rebase master later, and merge it if the CI is green.
